core/src/main/scala/org/apache/spark/SparkEnv.scala

```diff
@@ -355,6 +355,11 @@ object SparkEnv extends Logging {
       None
     }
 
+    // Set the application attemptId in the ExternalShuffleClient is applicable.
+    // If there is no attemptId assigned, set the attemptId to -1.
+    externalShuffleClient.foreach(
+      shuffleClient => shuffleClient.setAppAttemptId(conf.getInt(config.APP_ATTEMPT_ID.key, -1)))
```

Comment:
SparkEnv is created at L460:
https://github.com/apache/spark/blob/adbea252db0f61ed3aaa45ea20b704bad6c47408/core/src/main/scala/org/apache/spark/SparkContext.scala#L460

But application attemptId is set at L586 :
https://github.com/apache/spark/blob/adbea252db0f61ed3aaa45ea20b704bad6c47408/core/src/main/scala/org/apache/spark/SparkContext.scala#L586

So, isn't `conf.getInt(config.APP_ATTEMPT_ID.key, -1)` only return -1 here?

Reply:
Yes, we should not set it here - but along with existing `APP_ATTEMPT_ID` set.

Reply:
Driver case is fixed by directly calling the setAppAttemptId when setting the conf. Tested in our Yarn cluster, works fine in Yarn cluster mode.